### ISSUE TITLE
tfjs-tflite-node : undefined symbol TfLiteInterpreterGetInputTensor

### ISSUE THREAD

**Issue by bartbutenaers** [tensorflow/tfjs#6796]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Raspbian GNU/Linux 10 (buster)
- TensorFlow.js installed from (npm or script link): npm (tfjs-node)
- TensorFlow.js version: tfjs-node version 3.19.0

**Describe the problem**

When I try to load ***tfjs-tflite-node*** on my Raspberry Pi 4 (32 bit):
```
var tf = require('@tensorflow/tfjs-node');
var {loadTFLiteModel} = require('tfjs-tflite-node');
```
Then the second `require` statement fails:
>  Error: /home/pi/.node-red/node_modules/tfjs-tflite-node/build/Release/node_tflite_binding.node: undefined symbol: TfLiteInterpreterGetInputTensor

Hopefully somebody can give me some tips about thins I could try...

Thanks!!
Bart

**Comment by rthadur** [comment]
`var {loadTFLiteModel} = require('tfjs-tflite-node');
` I am not sure we have a package published for tfjs-tflite-node , can you please double check ?

**Comment by bartbutenaers** [comment]
@rthadur,
Thanks for the quick response!
I use [this ](https://www.npmjs.com/package/tfjs-tflite-node) package.  Is that what you mean?

**Comment by bartbutenaers** [comment]
I had upgraded my Raspberry 4 to 64 bit, but no luck.
The reason I upgraded it to 64 bit, is because I get the above error while trying to run [this](https://codelabs.developers.google.com/tensorflowjs-coral-tflite-node#1) codelab which mentions:

> For this codelab, we recommend using a Raspberry Pi running [Raspberry Pi OS (64-bit) with desktop](https://www.raspberrypi.com/software/operating-systems/#raspberry-pi-os-64-bit).

So my platform should be ok now, and the steps in the codelab are written ***very*** nicely by @mattsoulanille, 
Which means I must have forgotten something very stupid, but can't figure out what.

When I execute a `yarn build` of my tfjs-tflite-node, the log looks ok to me:
```
make: Entering directory '/home/pi/.node-red/node_modules/tfjs-tflite-node/build'
  CXX(target) Release/obj.target/node_tflite_binding/binding/node_tflite_binding.o
../binding/node_tflite_binding.cc: In member function ‘std::pair<Napi::Array, std::vector<tfjs_tflite_node::TensorInfo*> > tfjs_tflite_node::Interpreter::make_tensors(Napi::Env&, TfLiteInterpreter*, bool)’:
../binding/node_tflite_binding.cc:520:22: warning: ISO C++ says that these are ambiguous, even though the worst conversion for the first is better than the worst conversion for the second:
       tensor_array[id] = wrapped_tensor_info;
                      ^
In file included from /home/pi/.node-red/node_modules/tfjs-tflite-node/node_modules/node-addon-api/napi.h:2981,
                 from ../binding/node_tflite_binding.cc:19:
/home/pi/.node-red/node_modules/tfjs-tflite-node/node_modules/node-addon-api/napi-inl.h:1305:41: note: candidate 1: ‘Napi::Object::PropertyLValue<unsigned int> Napi::Object::operator[](uint32_t)’
 inline Object::PropertyLValue<uint32_t> Object::operator [](uint32_t index) {
                                         ^~~~~~
../binding/node_tflite_binding.cc:520:22: note: candidate 2: ‘operator[](napi_value {aka napi_value__*}, int)’ <built-in>
       tensor_array[id] = wrapped_tensor_info;
                      ^
  SOLINK_MODULE(target) Release/obj.target/node_tflite_binding.node
  COPY Release/node_tflite_binding.node
make: Leaving directory '/home/pi/.node-red/node_modules/tfjs-tflite-node/build'
gyp info ok
Done in 25.44s.
```
And when I search for "TFLiteInterpreterGetInputTensor", I see it is available in a series of files:
```
pi@raspberrypi:~/.node-red/node_modules/tfjs-tflite-node $ grep -rnw . -e "TfLiteInterpreterGetInputTensor"
./binding/node_tflite_binding.cc:504:      get_tensor = &TfLiteInterpreterGetInputTensor;
Binary file ./build/Release/obj.target/node_tflite_binding.node matches
Binary file ./build/Release/obj.target/node_tflite_binding/binding/node_tflite_binding.o matches
Binary file ./build/Release/node_tflite_binding.node matches
Binary file ./cc_deps/linux_arm64/libtensorflowlite_c.so matches
Binary file ./cc_deps/windows_amd64/tensorflowlite_c.dll.if.lib matches
Binary file ./cc_deps/windows_amd64/tensorflowlite_c.dll matches
./cc_deps/include/tensorflow/lite/c/c_api.h:52:///     TfLiteInterpreterGetInputTensor(interpreter, 0);
./cc_deps/include/tensorflow/lite/c/c_api.h:170:TFL_CAPI_EXPORT extern TfLiteTensor* TfLiteInterpreterGetInputTensor(
Binary file ./cc_deps/linux_amd64/libtensorflowlite_c.so matches
```
I am running out of creativity, so all tips are welcome ;-)

**Comment by mattsoulanille** [comment]
Hi @bartbutenaers. Thanks for the detailed log messages.

I've tried reproducing the error you're seeing. Here are the steps I followed (mostly from the codelab).
* Downloaded and flashed Raspbian `2022-04-04-raspios-bullseye-arm64` to a Raspberry Pi 4 Model B Rev 1.1 (although I've also tested this on a 3b+ a while ago)
* Downloaded [NodeJS 16.17.0](https://nodejs.org/dist/v16.17.0/) and added it to PATH.
* Cloned the codelab with `git clone https://github.com/tensorflow/sig-tfjs.git`
* Moved to `tfjs-tflite-node-codelab/cpu_inference_working` and ran `npm install`
* Ran the demo with `npm start`.

The demo works for me (I know, not really helpful). Maybe the issue you're running into has something to do with your Node version. Can you see if the above steps work and let me know what version of node-RED you're using? Thanks!

You could also try using `@tensorflow/tfjs` instead of `@tensorflow/tfjs-node`, although I don't think that's the issue.

**Comment by bartbutenaers** [comment]
@mattsoulanille,
Thanks for joining this discussion.  Much appreciated!!

1. I have now upgraded the RaspiOS (on my Raspberry Pi 4 Model B Rev 1.2) from buster to bullseye.
2. I have now upgraded Node.js  from version v14.20.0 to v16.17.0
   Npm is now version 8.15.0
4. I have cloned the sig-tfjs repo and runned the npm install on it
5. Then I have runned the npx yarn commands from your codelab.  

In the last step it fails at the third npx command (with the same error as I got in my Node-RED integration):
```
pi@raspberrypi:~/sig-tfjs/coral-tflite-delegate $ npx yarn test-integration
yarn run v1.22.19
warning ../../package.json: No license field
$ yarn build && yarn test-integration-dev
warning ../../package.json: No license field
$ tsc
warning ../../package.json: No license field
$ jasmine --config=jasmine-integration.json
Error: /home/pi/sig-tfjs/tfjs-tflite-node/build/Release/node_tflite_binding.node: undefined symbol: TfLiteInterpreterGetInputTensor
    at Object.Module._extensions..node (node:internal/modules/cjs/loader:1210:18)
    at Module.load (node:internal/modules/cjs/loader:1004:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
    at Module.require (node:internal/modules/cjs/loader:1028:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at bindings (/home/pi/sig-tfjs/tfjs-tflite-node/node_modules/bindings/bindings.js:112:48)
    at Object.<anonymous> (/home/pi/sig-tfjs/tfjs-tflite-node/dist/index.js:48:34)
    at Module._compile (node:internal/modules/cjs/loader:1126:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1180:10)
    at Module.load (node:internal/modules/cjs/loader:1004:32) {
  code: 'ERR_DLOPEN_FAILED'
}
error Command failed with exit code 1.
```
Will need to google a bit, because this is way out of my comfort zone to be honest...

**Comment by mattsoulanille** [comment]
It seems like the node binding that gets compiled by node-gyp during the build process is finding the `libtensorflow.so` library (you're not seeing `Error: libtensorflowlite_c.so: cannot open shared object file: No such file or directory`), but it's not finding the `TfLiteInterpreterGetInputTensor` symbol in that library.

```
node_tflite_binding.node: undefined symbol: TfLiteInterpreterGetInputTensor
```

Can you try running `npx yarn test` in `sig-tfjs/tfjs-tflite-node`? If this fails (likely), can you run `ldd build/Release/node_tflite_binding.node`? You should see something similar to this but for arm:
```
$ ldd build/Release/node_tflite_binding.node
	linux-vdso.so.1 (0x00007ffceefbf000)
	libtensorflowlite_c.so => /usr/local/google/home/msoulanille/sig-tfjs/tfjs-tflite-node/build/Release/../../cc_deps/linux_amd64/libtensorflowlite_c.so (0x00007f62e17cd000)
	libexternal_delegate_obj.so => /usr/local/google/home/msoulanille/sig-tfjs/tfjs-tflite-node/build/Release/../../cc_deps/linux_amd64/libexternal_delegate_obj.so (0x00007f62e17c8000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f62e1594000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f62e1574000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f62e1399000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f62e1393000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f62e1389000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f62e1246000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f62e1225000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f62e1b9d000)
```
Mostly, I'm interested in if  `libtensorflowlite_c.so`'s path contains `cc_deps/linux_arm64/libtensorflowlite_c.so`

You can also check if libtensorflowlite has the TfLiteInterpreterGetInputTensor symbol with the following command, but it should be impossible for it not to:
```
$ nm -DC cc_deps/linux_arm64/libtensorflowlite_c.so | grep TfLiteInterpreterGetInputTensor
000000000000fd18 T TfLiteInterpreterGetInputTensor@@VERS_1.0
000000000000fcf8 T TfLiteInterpreterGetInputTensorCount@@VERS_1.0
000000000000f588 T TfLiteInterpreterGetInputTensorIndex@@VERS_1.0
```

**Comment by bartbutenaers** [comment]
1. Running `npx yarn test` in `sig-tfjs/tfjs-tflite-node` indeed results in exactly the same stacktrace as above.

2. Running the `ldd` command results in the following shared objects:
   ```
   pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node $ ldd build/Release/node_tflite_binding.node
        /usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so => /usr/lib/arm-linux-gnueabihf/libarmmem-v8l.so (0xf7bd6000)
        libstdc++.so.6 => /lib/arm-linux-gnueabihf/libstdc++.so.6 (0xf7a7b000)
        libm.so.6 => /lib/arm-linux-gnueabihf/libm.so.6 (0xf79f9000)
        libgcc_s.so.1 => /lib/arm-linux-gnueabihf/libgcc_s.so.1 (0xf79cc000)
        libpthread.so.0 => /lib/arm-linux-gnueabihf/libpthread.so.0 (0xf79a2000)
        libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xf7855000)
        /lib/ld-linux-armhf.so.3 (0xf7c0e000)
   ```
   As you can see the *libtensorflowlite_c.so* is ***not*** in this list...

3. The `nm` fails to find symbols in that file:
   ```
   pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node/cc_deps/linux_arm64 $ ls
   libexternal_delegate_obj.so  libtensorflowlite_c.so

   pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node/cc_deps/linux_arm64 $ nm -DC libtensorflowlite_c.so | grep TfLiteInterpreterGetInputTensor
   nm: libtensorflowlite_c.so: file format not recognized
   ´´´

**Comment by mattsoulanille** [comment]
Hmm. It's strange that `nm` failed. Maybe there's something wrong with your copy of `libtensorflowlite_c.so`, although that should be impossible because it's tracked in Git.
```
$ file cc_deps/linux_arm64/libtensorflowlite_c.so 
cc_deps/linux_arm64/libtensorflowlite_c.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (GNU/Linux), dynamically linked, BuildID[md5/uuid]=43f8bc3134b9769fbe8c7a161f016c63, stripped
```
```
$ shasum -a 256 cc_deps/linux_arm64/*
7cc05a58e977a67b2907e60f43b3d9e18d9d5090a2292af3cf31b67645938b98  cc_deps/linux_arm64/libexternal_delegate_obj.so
ec0dee4fb3418e680d010248ac1e304d9656cc5c8de54caaf91d666712e2f633  cc_deps/linux_arm64/libtensorflowlite_c.so
```
Alternatively, there might be a file permissions issue, although this seems unlikely. I tried removing `x`, and it still worked. When I removed `r`, I got `Error: libtensorflowlite_c.so: cannot open shared object file: No such file or directory` as expected (no read permission).
```
$ ls -l cc_deps/linux_arm64/
total 3248
-rwxr-xr-x 1 msoulanille primarygroup   14304 Jun 30 16:34 libexternal_delegate_obj.so
-rwxr-xr-x 1 msoulanille primarygroup 3301592 Jun 30 16:34 libtensorflowlite_c.so
```
You can also try `objdump`, since `nm` is not working for some reason.
```
$ objdump -T cc_deps/linux_arm64/libtensorflowlite_c.so | grep TfLiteInterpreterGetInputTensor
000000000000fd18 g    DF .text	000000000000003c  VERS_1.0    TfLiteInterpreterGetInputTensor
000000000000f588 g    DF .text	0000000000000018  VERS_1.0    TfLiteInterpreterGetInputTensorIndex
000000000000fcf8 g    DF .text	000000000000001c  VERS_1.0    TfLiteInterpreterGetInputTensorCount
```
Or, as a last resort, `strings`
```
$ strings cc_deps/linux_arm64/libtensorflowlite_c.so | grep TfLiteInterpreterGetInputTensor
TfLiteInterpreterGetInputTensorIndex
TfLiteInterpreterGetInputTensorCount
TfLiteInterpreterGetInputTensor
```

Another possibility is there's something misconfigured in `binding.gyp` preventing it from linking properly. If your SHA256 is the same as mine, I'll look into this.

**Comment by bartbutenaers** [comment]
I like your approach to troubleshoot this kind of issues.  
Hopefully you don't need to spend too much of your time on it...

+ The `file` command output looks completely identical to yours:
   ```
   pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node/cc_deps/linux_arm64 $ file libtensorflowlite_c.so
   libtensorflowlite_c.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (GNU/Linux), dynamically linked, BuildID[md5/uuid]=43f8bc3134b9769fbe8c7a161f016c63, stripped

+ The `shasum` output also looks identical:
   ```
   pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node $ shasum -a 256 cc_deps/linux_arm64/*
   7cc05a58e977a67b2907e60f43b3d9e18d9d5090a2292af3cf31b67645938b98 cc_deps/linux_arm64/libexternal_delegate_obj.so
   ec0dee4fb3418e680d010248ac1e304d9656cc5c8de54caaf91d666712e2f633 cc_deps/linux_arm64/libtensorflowlite_c.so
   ```

+ The `ls` command shows me the same permissions, but in my case for owner and group ***"pi"***:
   ```
   pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node $ ls -l cc_deps/linux_arm64/
   total 3244
   -rwxr-xr-x 1 pi pi   14304 Aug 31 08:07 libexternal_delegate_obj.so
   -rwxr-xr-x 1 pi pi 3301592 Aug 31 08:07 libtensorflowlite_c.so
   ```
   So perhaps there is something different in the user or group permissions...

+ The `objdump` command fails:
   ```
   pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node $ objdump -T cc_deps/linux_arm64/libtensorflowlite_c.so | grep TfLiteInterpreterGetInputTensor
   objdump: cc_deps/linux_arm64/libtensorflowlite_c.so: file format not recognized
   ```

+ Finally the `string` command output looks like this:
   ```
   pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node $ strings cc_deps/linux_arm64/libtensorflowlite_c.so | grep 
   TfLiteInterpreterGetInputTensor
   TfLiteInterpreterGetInputTensorIndex
   TfLiteInterpreterGetInputTensorCount
   TfLiteInterpreterGetInputTensor
   ```

**Comment by terryheo** [comment]
According to your output at https://github.com/tensorflow/tfjs/issues/6796#issuecomment-1233381876,
you're using 32bit OS system (`gnueabihf`) instead of ARM 64bit (`aarch64`).
Could you check it again?
```
pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node $ ldd build/Release/node_tflite_binding.node
     /usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so => /usr/lib/arm-linux-gnueabihf/libarmmem-v8l.so (0xf7bd6000)
     libstdc++.so.6 => /lib/arm-linux-gnueabihf/libstdc++.so.6 (0xf7a7b000)
     libm.so.6 => /lib/arm-linux-gnueabihf/libm.so.6 (0xf79f9000)
     libgcc_s.so.1 => /lib/arm-linux-gnueabihf/libgcc_s.so.1 (0xf79cc000)
     libpthread.so.0 => /lib/arm-linux-gnueabihf/libpthread.so.0 (0xf79a2000)
     libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xf7855000)
     /lib/ld-linux-armhf.so.3 (0xf7c0e000)
```

**Comment by bartbutenaers** [comment]
Hi @terryheo,

It seems to be 64 bit to me:
```
pi@raspberrypi:~/sig-tfjs/tfjs-tflite-node $ uname -m
aarch64
```
It was 32-bit a few days ago, but then I upgraded to 64-bit via [these](https://www.tomshardware.com/how-to/upgrade-raspberry-pi-os-to-bullseye-from-buster) instructions.
Perhaps something is mixed up somehow, by upgrading from 32-bit to 64 bit. 
I will empty my SD card this weekend and start from scratch by installing a clean 64-bit Buster.
Will get back here with my feedback as soon as possible...
Thanks!!

**Comment by terryheo** [comment]
64 bit operating system supports 32bit user binaries.

Few years ago, RPI OS only supported 32bit, so I used to use Ubuntu to run 64bit binary.
https://ubuntu.com/download/raspberry-pi

The link you showed was OS upgrade to Bullseye from Buster. I don't think it's 64bit installation.
You might want to try this. https://www.makeuseof.com/install-64-bit-version-of-raspberry-pi-os/

**Comment by bartbutenaers** [comment]
@mattsoulanille: I can confirm that the `file` commands (both for ld and gcc) showed me that the tools were 32 bit.

@terryheo: I quickly had a look at installing a full 64 bit toolchain (gcc and so on...), but like you mentioned that seems not to be an easy task.

So at the end I have installed  Raspberry Pi OS 64 bit from scratch, using the Raspberry Pi Imager tool.  
Was not very tempted to do that, since it resulted in a lot of things that I needed to install and configure...

However I can confirm that the above problems are ***solved*** now!!
I can run the integration tests from the codelab without problems!
_**Summarized: it works fine if you have installed a full 64-bit OS.**_

But before I close this issue, it would be nice if somebody could explain me why this only works for 64 bit.
Not sure at the moment if that is a limitation of TfLite, or something else?
Or perhaps it is just because tfjs-tflite-node (which is still a work in progress) has been maintained until now only on 64 bit? 

It might be useful to mention this 64 bit limitation on some readme pages.

**Comment by mattsoulanille** [comment]
I'm glad it's working now, @bartbutenaers!

To answer your question about why 32 bit isn't supported yet, we just haven't had the time to add it yet. TFLite supports [32 bit arm](https://www.tensorflow.org/lite/guide/build_arm#cross-compilation_for_arm_with_cmake), but for the initial release, we chose a small set of platforms to make the build process simpler. In the future, we'll add more platforms, but this will likely come after we set up some form of build automation and move the binaries out of this git repository.

Thanks for pointing out the lack of supported platforms in the README. I'll add some better documentation for that.
